File: Java/cpp/Body.cpp

#include "Body.hpp"

#include <math.h>

using namespace std;

Body::Body() {
    this->rx    = -1;
    this->ry    = -1;
    this->rz    = -1;
    this->vx    = -1;
    this->vy    = -1;
    this->vz    = -1;
    this->mass  = -1;
}

Body::Body(double rx, double ry, double rz, double vx, double vy, double vz, double mass) {
    this->rx    = rx;
    this->ry    = ry;
    this->rz    = rz;
    this->vx    = vx;
    this->vy    = vy;
    this->vz    = vz;
    this->mass  = mass;
}

// updates at the next time step
void Body::update(double dt) {
    vx += dt * fx / mass;
    vy += dt * fy / mass;
    vz += dt * fz / mass;
    rx += dt * vx;
    ry += dt * vy;
    rz += dt * vz;
}

// returns distance between two 'bodies' (which
// could be multiple bodies in BH approximation)
double Body::distanceTo(Body* b) {
    double dx = rx - b->rx;
    double dy = ry - b->ry;
    double dz = rz - b->rz;
    return sqrt(dx*dx + dy*dy + dz*dz);
}

void Body::resetForce() {
    fx = 0.0;
    fy = 0.0;
    fz = 0.0;
}

// adds the forces together
void Body::addForce(Body* b) {
    double dx = b->rx - rx;
    double dy = b->ry - ry;
    double dz = b->rz - rz;
    double dist = sqrt(dx*dx + dy*dy + dz*dz + 1E-10);
    double F = (G * mass * b->mass) / (dist*dist);
    fx += F * dx / dist;
    fy += F * dy / dist;
    fz += F * dz / dist;
}
// checks if it's in the octant
bool Body::in(Oct* o) {
    return o->contains(rx, ry, rz);
}

// center of mass calculation
//TODO check if this is the correct formula
void Body::plus(Body* b) {
    Body* a = this;

    totalMass = a->mass + b->mass;
    a->rx = (a->rx * a->mass + b->rx * b->mass) / totalMass; //[(location * mass) + (location * mass)] / totalMass
    a->ry = (a->ry * a->mass + b->ry * b->mass) / totalMass;
    a->rz = (a->rz * a->mass + b->rz * b->mass) / totalMass;
    a->mass = totalMass;
    //works: ((((((2*7)+(3*5))/12) * 12) + (4*19))/31) == ((2*7)+(3*5)+(4*19))/31
}

void Body::copyFrom (Body* b){
    rx = b->rx;
    ry = b->ry;
    rz = b->rz;
    vx = b->vx;
    vy = b->vy;
    vz = b->vz;
    mass = b->mass;
}

File: Java/Body.java

public class Body {

    // gravitational constant
    private static final double G = 6.67e-11;

    private double rx, ry, rz;       // position
    private double vx, vy, vz;       // velocity
    private double fx, fy, fz;       // force
    private double mass;             // mass

    public Body(double rx, double ry, double rz, double vx, double vy, double vz, double mass) {
        this.rx    = rx;
        this.ry    = ry;
        this.rz    = rz;
        this.vx    = vx;
        this.vy    = vy;
        this.vz    = vz;
        this.mass  = mass;
    }

    // updates at the next time step
    public void update(double dt) {
        vx += dt * fx / mass;
        vy += dt * fy / mass;
        vz += dt * fz / mass;
        rx += dt * vx;
        ry += dt * vy;
        rz += dt * vz;
    }

    // returns distance between two 'bodies' (which
    // could be multiple bodies in BH approximation)
    public double distanceTo(Body b) {
        double dx = rx - b.rx;
        double dy = ry - b.ry;
        return Math.sqrt(dx*dx + dy*dy);
    }

    public void resetForce() {
        fx = 0.0;
        fy = 0.0;
        fz = 0.0;
    }

    // adds the forces together
    public void addForce(Body b) {
        Body a = this;
        double EPS = 3E4;      // softening parameter
        double dx = b.rx - a.rx;
        double dy = b.ry - a.ry;
        double dz = b.rz - a.rz;
        double dist = Math.sqrt(dx*dx + dy*dy + dz*dz);
        double F = (G * a.mass * b.mass) / (dist*dist + EPS*EPS);
        a.fx += F * dx / dist;
        a.fy += F * dy / dist;
        a.fz += F * dz / dist;
    }
    // Nice representation of a body
    public String toString() {
        return String.format("%10.3E %10.3E %10.3E %10.3E %10.3E %10.3E %10.3E", rx, ry, rz, vx, vy, vz, mass);
    }
    // checks if it's in the octant
    public boolean in(Oct o) {
        return o.contains(this.rx, this.ry, this.rz);
    }

    // center of mass calculation
    public Body plus(Body b) {
        Body a = this;

        double m = a.mass + b.mass;
        double x = (a.rx * a.mass + b.rx * b.mass) / m;
        double y = (a.ry * a.mass + b.ry * b.mass) / m;
        double z = (a.rz * a.mass + b.rz * b.mass) / m;

        return new Body(x, y, z, a.vx, b.vx, a.vy, m);
    }
}
